Feature: Mermaid pan & zoom

snopan opened the pull request:
This PR adds pan & zoom functionality in mermaid MD preview. There's a previous [PR](https://github.com/mjbvz/vscode-markdown-mermaid/pull/247) that attempted to add this functionality but had some issues. This PR tries address the issues mentioned in the comments.

https://github.com/mjbvz/vscode-markdown-mermaid/issues/125

<img width="640" alt="image" src="https://github.com/user-attachments/assets/d47ca593-916c-4cd9-98df-7e1838c452df">


The changes include:
* Rendering a button about the diagram when there are no errors in diagram
* When svg-pan-zoom hasn't initiated, the behavior should be same as before
* Only when the button is pressed, svg-pan-zoom is initiated allowing for pan and zoom. The sizing of diagram will be set to what it is at that point and it also starts recording any actions taken (pan/zoom) into ZoomStates
* When the markdown is re-rendered, it will go check ZoomStates and sync back to where it is allowing for zoom and editing at the same time
* Unrelated it also removes the `throw error` when a diagram fails to parse. From what I can see when mermaid library fails to parse the contents, we would catch the error and return a error message content. But for some reason it makes another throw error, so when you have multiple diagrams and one breaks it breaks everything else.

<img width="677" alt="image" src="https://github.com/user-attachments/assets/dafcfb30-5816-4c73-97ed-3ddac8bfd644">

after remove `throw error`
<img width="677" alt="image" src="https://github.com/user-attachments/assets/5066e900-4baa-43e8-95be-034fcee8e3af">

snopan commented:
Hey @mjbvz, can I get a review on this PR? Thanks!

mjbvz reviewed (commented):
Thanks for looking into this! Sorry it took me so long to test it

The panning and zooming seems to work well. However to start I think it should be presented differently. Let's do:

- Add a new vscode configuration setting that lets you enable/disable pan zoom entirely. It can default to on, but there needs to be a way to turn the feature off entirely

- Only show the pan/zoom controls when you hover over the diagram

- Potentially add a command or right click option on the diagram that overrides the VS Code configuration. So if I have pan zoom enabled in my settings, I'd like the command to be able to turn it off in case it's causing any issues

A few other points I noticed:

- The colors of the controls seem off, especially on dark themes. If you can, reuse some of the [vscode theme colors](https://code.visualstudio.com/api/extension-guides/webview#theming-webview-content). These should all be available inside of the markdown preview webview

- The diagram get cut off. I'd try to get it match what GitHub does instead where the diagram is centered and takes up more horizontal space

I will take a more detailed look once some of those points are addressed. Let me know if you have any questions about these

snopan commented:
Hey @mjbvz , thanks for getting back to me! I just slightly confused and want to clarify a few things:
* Do we still want to keep the toggle button or is the pan/zoom now controlled by this vscode configuration?
* So when you mention show pan/zoom controls on hover, do you mean the toggle button
<img width="202" alt="image" src="https://github.com/user-attachments/assets/64f78eac-8954-4f28-84ce-3b05dd392e78">

or these controls after pan/zoom is enabled
<img width="143" alt="image" src="https://github.com/user-attachments/assets/07137a90-95bf-4a4b-a5be-f32c6bbc94b3">
* When you say right click and turn off pan/zoom, does that also turn off pan/zoom entirely like the VS Configuration.
* Can you give me a bit more details or steps taken to get to this "diagram get cut off" behaviour? I mostly just rely on svg-pan-zoom library to do the pan/zoom and have some extra logic to store the state of pan/zoom.

mjbvz commented:
Sure, following up:

- The toggle feels too prominent to show on every single diagram. Let's start with using the configuration plus maybe command /right click for quick toggling just for the current preview

- For the controls, I mean the `+`/`-`. I only care about these if I'm interacting with the diagram. Again it feels cluttered to show them all the time

- Let's try to match how word wrap works in editors works. With word wrap, the setting provides the default value but you can then override this per editor using the command. The command however only applies to the current document, it does not change the configuration

- Try creating various sized diagram and then zooming in / panning. You can also try making the markdown preview itself larger or smaller. I was fairly easily able to get the diagram to be cut off horizontally when doing this

snopan commented:
Hey @mjbvz, sorry haven't been able to work on it much been kinda busy with other stuff. 

I've made a few changes according to some of your comments

### Toggle and Controls being too prominent
* Added a vscode config field "enablePanZoom" which by default is true but when set to false will not render with any pan zoom functionality
* Both toggle and controls will only appear when hovered on so they aren't there all the time
* Given the above two changes, I didn't work on adding any commands or right click functionality, feels like the code would get a bit more messier and need to refactor a tad bit.

### Color theme issues
* For the toggle I've changed the styling to use vscode color variables, not sure if they're the exact correct ones. I just picked a background that is darker than the preview background then just a foreground and active color.
* For the Controls that one is a bit more tricky, so that is created by svg-pan-zoom and a bit difficult to edit the styling. The way to get around it is to disable their controls and we create our own custom one, which is a bit more work. Not sure if we can put it as an uplift item later on.

### Diagram being cut off
So for this issue I'm assuming you were talking about the diagram not taking the full width of the page when pan zoom is enabled. The reason for this was because I wanted the switching between pan zoom to look seamless, so when it gets turned on it takes the current size dimension of the diagram. But given it doesn't use the full width of the page would look like it's cut off.
![image](https://github.com/user-attachments/assets/fc907a99-59c7-4115-8b90-10fb9d39eaae)

I've updated so that it does take the full width of the page when pan zoom is enabled, but the height is still the height of the current diagram so when you resize the width down it does look a bit skinny and tall which is something that we have to accept I think because the diagram needs a set height when pan zoom is enabled. 
![image](https://github.com/user-attachments/assets/01cda36a-2f1b-4a67-a34f-52759e7e51eb)
![image](https://github.com/user-attachments/assets/c64dc732-d5f5-469f-be19-06be375bd2c3)

snopan commented:
@mjbvz just a bump on this PR, if you're free to have a look that would be great!

jmwachtel commented:
@snopan Thank you for starting this, I am equally interested in this functionality. @mjbvz do you think you can help move this along?

OdyAsh commented:
> @snopan Thank you for starting this, I am equally interested in this functionality.

+1 🙌

ioggstream commented:
@snopan if you don't have time to rebase, I did it here: https://github.com/ioggstream/vscode-d3fend-editor/tree/bierner-snopan
